max-lines-per-function ignores arrow IIFEs

**Tell us about your environment**
* **ESLint Version:** 7.0.0
* **Node Version:** 12.16.2
* **npm Version:** 6.14.4

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**
<details>
<summary>Configuration</summary>

```json
{
  "env": {
    "es6": true
  },
  "rules": {
    "max-lines-per-function": [2, 1]
  }
}

```
</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
(() => {
})()
```

```bash
eslint test.js
```

[Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG1heC1saW5lcy1wZXItZnVuY3Rpb246IFsyLCAxXSAqL1xuKCgpID0+IHtcbn0pKCkiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjYsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=)

**What did you expect to happen?**
No errors because `max-lines-per-function` does not include any code in IIFEs by default.

**What actually happened? Please include the actual, raw output from ESLint.**
```
/path/to/test.js
  1:2  error  Arrow function has too many lines (2). Maximum allowed is 1  max-lines-per-function

✖ 1 problem (1 error, 0 warnings)
```

**Are you willing to submit a pull request to fix this bug?**
Yes.

[lib/rules/max-lines-per-function.js:137](https://github.com/eslint/eslint/blob/master/lib/rules/max-lines-per-function.js#L137)
```diff
-            return node.type === "FunctionExpression" && node.parent && node.parent.type === "CallExpression" && node.parent.callee === node;
+            return (node.type === "FunctionExpression" || node.type === "ArrowFunctionExpression") && node.parent && node.parent.type === "CallExpression" && node.parent.callee === node;
```

--- a/docs/rules/max-lines-per-function.md
+++ b/docs/rules/max-lines-per-function.md
@@ -165,6 +165,10 @@ Examples of **incorrect** code for this rule with the `{ "IIFEs": true }` option
 (function(){
     var x = 0;
 }());
+
+(() => {
+    var x = 0;
+})();
 ```
 
 Examples of **correct** code for this rule with the `{ "IIFEs": true }` option:
@@ -174,6 +178,10 @@ Examples of **correct** code for this rule with the `{ "IIFEs": true }` option:
 (function(){
     var x = 0;
 }());
+
+(() => {
+    var x = 0;
+})();
 ```
 
 ## When Not To Use It

--- a/lib/rules/max-lines-per-function.js
+++ b/lib/rules/max-lines-per-function.js
@@ -134,7 +134,7 @@ module.exports = {
          * @returns {boolean} True if it's an IIFE
          */
         function isIIFE(node) {
-            return node.type === "FunctionExpression" && node.parent && node.parent.type === "CallExpression" && node.parent.callee === node;
+            return (node.type === "FunctionExpression" || node.type === "ArrowFunctionExpression") && node.parent && node.parent.type === "CallExpression" && node.parent.callee === node;
         }
 
         /**

--- a/tests/lib/rules/max-lines-per-function.js
+++ b/tests/lib/rules/max-lines-per-function.js
@@ -164,6 +164,30 @@ if ( x === y ) {
     return bar;
 }());`,
             options: [{ max: 2, skipComments: true, skipBlankLines: false, IIFEs: false }]
+        },
+
+        // Arrow IIFEs should be recognised if IIFEs: true
+        {
+            code: `(() => {
+    let x = 0;
+    let y = 0;
+    let z = x + y;
+    let foo = {};
+    return bar;
+})();`,
+            options: [{ max: 7, skipComments: true, skipBlankLines: false, IIFEs: true }]
+        },
+
+        // Arrow IIFEs should not be recognised if IIFEs: false
+        {
+            code: `(() => {
+    let x = 0;
+    let y = 0;
+    let z = x + y;
+    let foo = {};
+    return bar;
+})();`,
+            options: [{ max: 2, skipComments: true, skipBlankLines: false, IIFEs: false }]
         }
     ],
 
@@ -434,6 +458,21 @@ if ( x === y ) {
             options: [{ max: 2, skipComments: true, skipBlankLines: false, IIFEs: true }],
             errors: [
                 { messageId: "exceed", data: { name: "Function", lineCount: 7, maxLines: 2 } }
+            ]
+        },
+
+        // Test the IIFEs option includes arrow IIFEs
+        {
+            code: `(() => {
+    let x = 0;
+    let y = 0;
+    let z = x + y;
+    let foo = {};
+    return bar;
+})();`,
+            options: [{ max: 2, skipComments: true, skipBlankLines: false, IIFEs: true }],
+            errors: [
+                { messageId: "exceed", data: { name: "Arrow function", lineCount: 7, maxLines: 2 } }
             ]
         }
     ]

